[WebGPU] Virtualize PresentationContext

litherum opened the pull request:
#### 6c91a810e7ee70363e0f9e9a888ed7f29837a841
<pre>
[WebGPU] Virtualize PresentationContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=250632">https://bugs.webkit.org/show_bug.cgi?id=250632</a>
rdar://104270153

Reviewed by Tadeu Zagallo.

PresentationContext is how we composite results into a window. Its implementation
in <a href="https://bugs.webkit.org/show_bug.cgi?id=247587">https://bugs.webkit.org/show_bug.cgi?id=247587</a> is set up only for WebKit&apos;s compositing
using IOSurfaces. However, it&apos;s convenient if we could have another implementation of them
that could use CAMetalLayer directly, so that we could develop WebGPU without having to build
all of WebCore.

This patch turns PresentationContext into a base class with virtual methods, and provides
two implementations of them: One for IOSurfaces, that WebKit will use, and one for CAMetalLayer.
The IOSurface implementation isn&apos;t modified at all; it&apos;s simply moved from the existing
implementation.

This patch doesn&apos;t actually implement the CAMetalLayer implementation; this patch is already
quite large without it, and I want to keep my patches small to aid reviewing. I will implement
getCurrentTextureView() and present() in my next patch.

No test because there is no behavior change.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/PresentationContext.h:
(WebGPU::PresentationContext::createInvalid):
(WebGPU::PresentationContext::isPresentationContextIOSurface const):
(WebGPU::PresentationContext::isPresentationContextCoreAnimation const):
(WebGPU::PresentationContext::create): Deleted.
(WebGPU::PresentationContext::displayBuffer const): Deleted.
(WebGPU::PresentationContext::drawingBuffer const): Deleted.
* Source/WebGPU/WebGPU/PresentationContext.mm:
(WebGPU::PresentationContext::create):
(WebGPU::PresentationContext::getPreferredFormat):
(WebGPU::PresentationContext::present):
(WebGPU::optionsFor32BitSurface): Deleted.
(WebGPU::createIOSurface): Deleted.
(WebGPU::createSurfaceFromDescriptor): Deleted.
(WebGPU::Device::createSurface): Deleted.
(WebGPU::PresentationContext::PresentationContext): Deleted.
(WebGPU::PresentationContext::nextDrawable): Deleted.
(wgpuSurfaceCocoaCustomSurfaceGetDisplayBuffer): Deleted.
(wgpuSurfaceCocoaCustomSurfaceGetDrawingBuffer): Deleted.
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.h: Copied from Source/WebGPU/WebGPU/PresentationContext.h.
(WebGPU::PresentationContextCoreAnimation::create):
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.mm: Copied from Source/WebGPU/WebGPU/PresentationContext.h.
(WebGPU::PresentationContextCoreAnimation::PresentationContextCoreAnimation):
(WebGPU::PresentationContextCoreAnimation::configure):
(WebGPU::PresentationContextCoreAnimation::present):
(WebGPU::PresentationContextCoreAnimation::getCurrentTextureView):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.h: Copied from Source/WebGPU/WebGPU/PresentationContext.h.
(WebGPU::PresentationContextIOSurface::create):
(WebGPU::PresentationContextIOSurface::displayBuffer const):
(WebGPU::PresentationContextIOSurface::drawingBuffer const):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm: Copied from Source/WebGPU/WebGPU/PresentationContext.mm.
(WebGPU::optionsFor32BitSurface):
(WebGPU::createIOSurface):
(WebGPU::createSurfaceFromDescriptor):
(WebGPU::PresentationContextIOSurface::PresentationContextIOSurface):
(WebGPU::PresentationContextIOSurface::configure):
(WebGPU::PresentationContextIOSurface::present):
(WebGPU::PresentationContextIOSurface::getCurrentTextureView):
(WebGPU::PresentationContextIOSurface::nextDrawable):
(wgpuSurfaceCocoaCustomSurfaceGetDisplayBuffer):
(wgpuSurfaceCocoaCustomSurfaceGetDrawingBuffer):

Canonical link: <a href="https://commits.webkit.org/259022@main">https://commits.webkit.org/259022@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8cccfeb113911d3872009ff5e7be4e02e349181d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103692 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12808 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36638 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112916 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13830 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3703 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95928 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/112064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109463 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93723 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/38384 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92484 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25329 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80034 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6171 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/26722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6346 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3250 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12327 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46244 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6198 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8105 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/SurfaceWebKit.h:
```diff
@@ -0,0 +1,60 @@
+
+namespace WebGPU {
+
+class SurfaceWebKit : public Surface {
```
Is `SurfaceWebKit` the correct name? Since it is simply backed by an IOSurface. A non-webkit test app could use `SurfaceWebKit` without using WebKit if they used IOSurfaces. `SurfaceIOSurface` doesn't sound right either. Maybe `NativeSurface` and rename `SurfaceCoreAnimation` to `MetalSurface`? I can think about it more too.

This is a minor detail as we can rename later easily. I am also ok with SurfaceWebKit for now. Especially since this is created via `wgpuInstanceCreateSurface` the name is even less important

litherum replied:
Renamed to PresentationContextIOSurface.

mwyrzykowski commented on Source/WebGPU/WebGPU/SwapChainWebKit.h:
```diff
@@ -0,0 +1,54 @@
+    SwapChainWebKit(SurfaceWebKit&, const WGPUSwapChainDescriptor&);
+
+    Ref<SurfaceWebKit> m_surface;
+    RetainPtr<IOSurfaceRef> m_drawingBuffer;
```
I think we can remove this variable

litherum replied:
Done.

litherum commented on Source/WebGPU/WebGPU/Surface.h:
```diff
@@ -29,36 +29,42 @@
-private:
-    Surface(const WGPUSurfaceDescriptor&);
-    Surface(int, int);
+    virtual Ref<SwapChain> createSwapChain(const Device&, const WGPUSwapChainDescriptor&);
```
Device should not be const.

litherum replied:
Done.

mwyrzykowski commented on Source/WebGPU/WebGPU/SwapChainCoreAnimation.h:
```diff
@@ -0,0 +1,55 @@
+private:
+    SwapChainCoreAnimation(const Device&, const WGPUSwapChainDescriptor&, CAMetalLayer *);
+
+    CAMetalLayer *m_layer { nil };
```
Similar to the comment https://github.com/WebKit/WebKit/pull/8665/files#r1070535343 I think CAMetalLayer should be in SurfaceCoreAnimation and not here

djg replied:
I disagree. CAMetalLayer is the SwapChain, isn't it?

litherum replied:
This was done earlier.

djg commented on Source/WebGPU/WebGPU/Surface.mm:
```diff
@@ -28,108 +28,47 @@
 }
 
-RetainPtr<IOSurfaceRef> Surface::nextDrawable()
+Ref<SwapChain> Surface::createSwapChain(const Device&, const WGPUSwapChainDescriptor&)
```
Why does a Surface create a SwapChain? Isn't a SwapChain a FIFO of Surfaces?

mwyrzykowski commented on Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
```diff
@@ -0,0 +1,137 @@
+
+void PresentationContextIOSurface::present()
+{
+    nextDrawable();
```
this has become more of like `swapBuffers` instead of `nextDrawable` because we ignore the return value, but I think changing the name is outside the scope of this PR, we can address it in the future

litherum replied:
Should I inline the contents of `PresentationContextIOSurface::nextDrawable()`? It's only called from one place.

litherum replied:
I'll keep it the way it is for now; I think we've got some larger refactoring to do here anyway in the near future.

mwyrzykowski replied:
Sounds good 👍, seems reasonable to inline it at this point but certainly no rush to do so

litherum committed:
[WebGPU] Virtualize PresentationContext
https://bugs.webkit.org/show_bug.cgi?id=250632
rdar://104270153

Reviewed by Tadeu Zagallo.

PresentationContext is how we composite results into a window. Its implementation
in https://bugs.webkit.org/show_bug.cgi?id=247587 is set up only for WebKit's compositing
using IOSurfaces. However, it's convenient if we could have another implementation of them
that could use CAMetalLayer directly, so that we could develop WebGPU without having to build
all of WebCore.

This patch turns PresentationContext into a base class with virtual methods, and provides
two implementations of them: One for IOSurfaces, that WebKit will use, and one for CAMetalLayer.
The IOSurface implementation isn't modified at all; it's simply moved from the existing
implementation.

This patch doesn't actually implement the CAMetalLayer implementation; this patch is already
quite large without it, and I want to keep my patches small to aid reviewing. I will implement
getCurrentTextureView() and present() in my next patch.

No test because there is no behavior change.

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/PresentationContext.h:
(WebGPU::PresentationContext::createInvalid):
(WebGPU::PresentationContext::isPresentationContextIOSurface const):
(WebGPU::PresentationContext::isPresentationContextCoreAnimation const):
(WebGPU::PresentationContext::create): Deleted.
(WebGPU::PresentationContext::displayBuffer const): Deleted.
(WebGPU::PresentationContext::drawingBuffer const): Deleted.
* Source/WebGPU/WebGPU/PresentationContext.mm:
(WebGPU::PresentationContext::create):
(WebGPU::PresentationContext::getPreferredFormat):
(WebGPU::PresentationContext::present):
(WebGPU::optionsFor32BitSurface): Deleted.
(WebGPU::createIOSurface): Deleted.
(WebGPU::createSurfaceFromDescriptor): Deleted.
(WebGPU::Device::createSurface): Deleted.
(WebGPU::PresentationContext::PresentationContext): Deleted.
(WebGPU::PresentationContext::nextDrawable): Deleted.
(wgpuSurfaceCocoaCustomSurfaceGetDisplayBuffer): Deleted.
(wgpuSurfaceCocoaCustomSurfaceGetDrawingBuffer): Deleted.
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.h: Copied from Source/WebGPU/WebGPU/PresentationContext.h.
(WebGPU::PresentationContextCoreAnimation::create):
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.mm: Copied from Source/WebGPU/WebGPU/PresentationContext.h.
(WebGPU::PresentationContextCoreAnimation::PresentationContextCoreAnimation):
(WebGPU::PresentationContextCoreAnimation::configure):
(WebGPU::PresentationContextCoreAnimation::present):
(WebGPU::PresentationContextCoreAnimation::getCurrentTextureView):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.h: Copied from Source/WebGPU/WebGPU/PresentationContext.h.
(WebGPU::PresentationContextIOSurface::create):
(WebGPU::PresentationContextIOSurface::displayBuffer const):
(WebGPU::PresentationContextIOSurface::drawingBuffer const):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm: Copied from Source/WebGPU/WebGPU/PresentationContext.mm.
(WebGPU::optionsFor32BitSurface):
(WebGPU::createIOSurface):
(WebGPU::createSurfaceFromDescriptor):
(WebGPU::PresentationContextIOSurface::PresentationContextIOSurface):
(WebGPU::PresentationContextIOSurface::configure):
(WebGPU::PresentationContextIOSurface::present):
(WebGPU::PresentationContextIOSurface::getCurrentTextureView):
(WebGPU::PresentationContextIOSurface::nextDrawable):
(wgpuSurfaceCocoaCustomSurfaceGetDisplayBuffer):
(wgpuSurfaceCocoaCustomSurfaceGetDrawingBuffer):

Canonical link: https://commits.webkit.org/259022@main
